plugins/libplugin-pay.c

```diff
@@ -356,16 +447,89 @@ static struct command_result *
 {
+	struct payment *root;
+	struct channel_hint hint;
+	struct route_hop *hop;
+	assert(p->route != NULL);
+
 	p->result = tal_sendpay_result_from_json(p, buffer, toks);
 
-	if (p->result->state == PAYMENT_COMPLETE)
+	if (p->result->state == PAYMENT_COMPLETE) {
 		p->step = PAYMENT_STEP_SUCCESS;
-	else
-		p->step = PAYMENT_STEP_FAILED;
+		goto cont;
+	}
 
-	/* TODO examine the failure and eventually stash exclusions that we
-	 * learned in the payment, so sub-payments can avoid them. */
+	p->step = PAYMENT_STEP_FAILED;
+	root = payment_root(p);
+
+	switch (p->result->failcode) {
+	case WIRE_PERMANENT_CHANNEL_FAILURE:
+	case WIRE_CHANNEL_DISABLED:
+	case WIRE_UNKNOWN_NEXT_PEER:
+	case WIRE_REQUIRED_CHANNEL_FEATURE_MISSING:
+		/* All of these result in the channel being marked as disabled. */
+		assert(*p->result->erring_index < tal_count(p->route));
+		hop = &p->route[*p->result->erring_index];
+		hint.enabled = false;
+		hint.scid.scid = hop->channel_id;
+		hint.scid.dir = hop->direction;
+		hint.estimated_capacity = AMOUNT_MSAT(0);
+		tal_arr_expand(&root->channel_hints, hint);
+		break;
+	case WIRE_TEMPORARY_CHANNEL_FAILURE:
+		/* These are an indication that the capacity was insufficient,
+		 * remember the amount we tried as an estimate. */
+		assert(*p->result->erring_index < tal_count(p->route));
+		hop = &p->route[*p->result->erring_index];
+		hint.enabled = true;
+		hint.scid.scid = hop->channel_id;
+		hint.scid.dir = hop->direction;
+		hint.estimated_capacity.millisatoshis = hop->amount.millisatoshis * 0.75; /* Raw: Multiplication */
+		tal_arr_expand(&root->channel_hints, hint);
+		break;
+
+	case WIRE_INVALID_ONION_PAYLOAD:
+	case WIRE_INVALID_REALM:
+	case WIRE_PERMANENT_NODE_FAILURE:
+	case WIRE_TEMPORARY_NODE_FAILURE:
+	case WIRE_REQUIRED_NODE_FEATURE_MISSING:
+	case WIRE_INVALID_ONION_VERSION:
+	case WIRE_INVALID_ONION_HMAC:
+	case WIRE_INVALID_ONION_KEY:
+#if EXPERIMENTAL_FEATURES
+	case WIRE_INVALID_ONION_BLINDING:
+#endif
+		/* These are reported by the last hop, i.e., the destination of hop i-1. */
+		assert(*p->result->erring_index - 1 < tal_count(p->route));
+		hop = &p->route[*p->result->erring_index - 1];
+		tal_arr_expand(&root->excluded_nodes, hop->nodeid);
+		break;
+
+ 	case WIRE_INCORRECT_OR_UNKNOWN_PAYMENT_DETAILS:
+	case WIRE_MPP_TIMEOUT:
+		/* These are permanent failures that should abort all of our
+		 * attempts right away. We'll still track pending partial
+		 * payments correctly, just not start new ones. */
+		root->abort = true;
+		break;
 
+	case WIRE_AMOUNT_BELOW_MINIMUM:
+	case WIRE_EXPIRY_TOO_FAR:
+	case WIRE_EXPIRY_TOO_SOON:
+	case WIRE_FEE_INSUFFICIENT:
+	case WIRE_INCORRECT_CLTV_EXPIRY:
+	case WIRE_FINAL_INCORRECT_CLTV_EXPIRY:
+		/* These are issues that are due to gossipd being out of date,
+		 * we ignore them here, and wait for gossipd to adjust
+		 * instead. */
+		break;
+	case WIRE_FINAL_INCORRECT_HTLC_AMOUNT:
+		/* These are symptoms of intermediate hops tampering with the
+		 * payment. */
```

Comment:
should we log these as unusual or info?

Reply:
We could split them out, but we need to react to it automatically anyway.

Reply:
right, my reason for suggesting to log them is that they might be information of interest to the node operator -- not immediately actionable but maybe something you'd want to have notes about

Reply:
Added logging message for this case.